airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/S3StorageOperations.java

```diff
@@ -43,21 +46,20 @@ public S3StorageOperations(final NamingConventionTransformer nameTransformer, fi
   }
 
   @Override
-  public String getBucketObjectName(final String namespace, final String streamName) {
-    return nameTransformer.applyDefaultCase(String.join("_",
-        nameTransformer.convertStreamName(namespace),
-        nameTransformer.convertStreamName(streamName)));
-  }
-
-  @Override
-  public String getBucketObjectPath(final String prefix, final String namespace, final String streamName, final DateTime writeDatetime) {
-    return nameTransformer.applyDefaultCase(String.format("%s/%s/%s/%02d/%02d/%02d/",
-        prefix,
-        getBucketObjectName(namespace, streamName),
-        writeDatetime.year().get(),
-        writeDatetime.monthOfYear().get(),
-        writeDatetime.dayOfMonth().get(),
-        writeDatetime.hourOfDay().get()));
+  public String getBucketObjectPath(final String namespace, final String streamName, final DateTime writeDatetime, final String customPathFormat) {
+    final String namespaceStr = nameTransformer.convertStreamName(isNotBlank(namespace) ? namespace : "");
+    final String streamNameStr = nameTransformer.convertStreamName(streamName);
+    return nameTransformer.applyDefaultCase(
+        customPathFormat
+            .replaceAll(Pattern.quote("${NAMESPACE}"), namespaceStr)
+            .replaceAll(Pattern.quote("${STREAM_NAME}"), streamNameStr)
```

Comment:
at the moment, the format is hard-coded to default, but I've seen this ask, should we also expose the format string in the spec of the connector?

Reply:
i added it to the spec.json

Reply:
yeah this seems pretty reasonable. Regardless, the equivalent of "bucket object path" is still connector-specific, so it would still need to happen per-connector?

Reply:
destination-s3's bucket objectPath ends up resolving as `s3://<bucket_name>/<namespace>/<prefix><stream_name>/<uuid>.csv.gz`

So with the current UI, it is possible to tweak <namespace> and <prefix> in the connection page, but not the other fields.
(<bucket_name> is configured on the destination settings page)